File: tests/fixtures/sample_codebase/router.py
"""
HTTP routing and request handling.
"""
from typing import Callable, Dict, Any


class Request:
    """Represents an HTTP request."""
    
    def __init__(self, path: str, method: str, headers: dict):
        self.path = path
        self.method = method
        self.headers = headers
        self.body = None


class Response:
    """Represents an HTTP response."""
    
    def __init__(self, status: int, body: Any, headers: dict = None):
        self.status = status
        self.body = body
        self.headers = headers or {}


class Router:
    """
    Main HTTP router.
    Routes requests to registered handlers.
    """
    
    def __init__(self, auth_handler):
        """
        Initialize router with authentication.
        
        Args:
            auth_handler: Auth instance for token validation
        """
        self.auth_handler = auth_handler
        self.routes: Dict[str, Callable] = {}
    
    def register(self, path: str, handler: Callable) -> None:
        """
        Register a handler for a path.
        
        Args:
            path: URL path pattern
            handler: Callable that handles the request
        """
        self.routes[path] = handler
    
    def route(self, request: Request) -> Response:
        """
        Route a request to the appropriate handler.
        
        Args:
            request: HTTP request to route
        
        Returns:
            HTTP response from handler
        """
        # Check authentication if auth token is present
        auth_token = request.headers.get('Authorization', '')
        if auth_token:
            if not self._validate_auth(auth_token):
                return Response(401, {'error': 'Unauthorized'})
        
        # Find handler
        handler = self.routes.get(request.path)
        if not handler:
            return self._not_found(request)
        
        # Execute handler
        try:
            return handler(request)
        except Exception as e:
            return self._error_response(e)
    
    def _validate_auth(self, token: str) -> bool:
        """
        Validate authentication token.
        
        Args:
            token: JWT token from request
        
        Returns:
            True if valid
        """
        try:
            # Strip 'Bearer ' prefix if present
            if token.startswith('Bearer '):
                token = token[7:]
            return self.auth_handler.validate_token(token)
        except ValueError:
            return False
    
    def _not_found(self, request: Request) -> Response:
        """Generate 404 response."""
        return Response(404, {
            'error': 'Not found',
            'path': request.path
        })
    
    def _error_response(self, error: Exception) -> Response:
        """Generate error response."""
        return Response(500, {
            'error': 'Internal server error',
            'message': str(error)
        })

File: tests/fixtures/sample_js_codebase/router.js
/**
 * HTTP Router — maps URL paths to handler functions.
 * Integrates with Auth for protected route validation.
 */

const { Auth } = require('./auth');

/**
 * Simple path-based HTTP router with middleware support.
 */
class Router {
    constructor() {
        this._routes = new Map();
        this._auth = null;
    }

    /**
     * Attach an Auth instance for protected routes.
     * @param {Auth} auth - Configured Auth instance.
     */
    useAuth(auth) {
        this._auth = auth;
    }

    /**
     * Register a handler for a given path.
     * @param {string} path - URL path pattern (e.g. '/api/users').
     * @param {Function} handler - Request handler (req, res) => void.
     * @param {Object} [options] - Options: { protected: boolean }.
     */
    register(path, handler, options = {}) {
        this._routes.set(path, { handler, options });
    }

    /**
     * Route a request to the appropriate handler.
     * Returns a response-like object: { status, body }.
     * @param {Object} request - { path, method, headers, body }.
     * @returns {{ status: number, body: any }}
     */
    route(request) {
        const route = this._routes.get(request.path);
        if (!route) {
            return { status: 404, body: { error: 'Not found' } };
        }

        if (route.options.protected) {
            const authed = this._validateAuth(request);
            if (!authed) {
                return { status: 401, body: { error: 'Unauthorized' } };
            }
        }

        try {
            return route.handler(request);
        } catch (err) {
            return { status: 500, body: { error: err.message } };
        }
    }

    // ── Private ────────────────────────────────────────────────────────────

    /**
     * Validate the Bearer token in request headers.
     * @param {Object} request
     * @returns {boolean}
     */
    _validateAuth(request) {
        if (!this._auth) return false;
        const authHeader = (request.headers || {}).authorization || '';
        const token = authHeader.replace(/^Bearer\s+/i, '');
        return this._auth.validateToken(token);
    }
}

module.exports = { Router };
